app/models/form_repository_rows_field_value.rb

```diff
@@ -62,12 +74,83 @@ def formatted
 
   def repository_row_snapshots(repository_row_ids)
     RepositoryRow.includes(repository_cells: %i(repository_column value)).where(id: repository_row_ids).map do |repository_row|
-      repository_row.as_json(include: { repository_cells: { include: %i(repository_column value) } }).merge(
+      repository_row.as_json(include: { repository_cells: { include: [:repository_column, :value, RELATIONAL_VALUE_INCLUDES] } }).merge(
         {
           'snapshot_at' => DateTime.current,
           'snapshot_by_id' => created_by_id
         }
       )
     end
   end
+
+  def reified_repository_cell(repository_row, cell_attributes)
+    repository_cell = repository_row.repository_cells.build(
+      cell_attributes.except('value', 'repository_column', *RELATIONAL_VALUE_INCLUDES.keys.map(&:to_s))
+    )
+
+    repository_column = RepositoryColumn.new(cell_attributes['repository_column'])
+    repository_column.readonly!
+
+    repository_cell.assign_attributes(
+      repository_column_id: cell_attributes['repository_column']['id'],
+      repository_row_id: repository_row.id,
+      repository_column: repository_column,
+      value: reified_repository_cell_value(repository_row, cell_attributes)
+    )
+
+    repository_cell.readonly!
+    repository_cell
+  end
+
+  def reified_repository_cell_value(repository_row, cell_attributes)
+    cell_value = cell_attributes['repository_column']['data_type'].constantize.new(
+      cell_attributes['value'].except('repository_column')
+    )
+
+    # mock relational values in-memory
+    case cell_value
+    when RepositoryAssetValue
+      snapshot_file = snapshot_files.find do |f|
+        f.blob.id == cell_attributes.dig('repository_asset_value', 'asset', 'file', 'blob', 'id')
+      end
+
+      cell_value.asset = Asset.new(
+        cell_attributes['repository_asset_value']['asset'].except('file')
+      )
+
+      cell_value.asset.file = snapshot_file.blob
+
+      cell_value.asset.readonly!
+    when RepositoryListValue
+      cell_value.define_singleton_method(:repository_list_item) do
```

Comment:
Why we can not directly assign a new item to the value?

Reply:
Yes, here it's not necessary since it's a direct association, will fix.